include/tvm/compilation.h

```diff
@@ -0,0 +1,211 @@
+    DLDeviceType deviceType;
+    /*! \brief The maximum threads that a schedule should use for this device */
+    int max_num_threads;
+    /*! \brief The warp size that should be used by the LowerThreadAllreduce pass */
+    int thread_warp_size;
+    /*! \brief Keys for this target */
+    std::unordered_set<std::string> keys;
+    /*! \brief Options for this target */
+    std::unordered_set<std::string> options;
+
+
+    Target(std::string targetName, DLDeviceType deviceType, int max_num_threads,
+        int thread_warp_size, const std::unordered_set<std::string>& keys,
+        const std::unordered_set<std::string>& options) {
+        this->targetName = targetName;
+        this->deviceType = deviceType;
+        this->max_num_threads = max_num_threads;
+        this->thread_warp_size = thread_warp_size;
+        this->keys = keys;
+        this->options = options;
+    }
+
+    /*! \return the full device string to pass to codegen::Build */
+    std::string str() const {
+        std::ostringstream result;
+        result << targetName;
+        for (const auto &x : options) {
+            result << " " << x;
+        }
+        return result.str();
+    }
+
+    /*! \return True iff this target has the given key */
+    bool hasKey(std::string key) const {
+        return keys.count(key) > 0;
+    }
+};
+
+/*! \return A target for LLVM */
+Target target_llvm() {
+    std::unordered_set<std::string> keys({ "llvm", "cpu" });
+    std::unordered_set<std::string> options;
+    return Target("llvm", kDLCPU, 512, 1, keys, options);
+}
+
+/*! \return A target for CUDA */
+Target target_cuda() {
+    std::unordered_set<std::string> keys({ "cuda", "gpu" });
+    std::unordered_set<std::string> options;
+    return Target("cuda", kDLGPU, 512, 32, keys, options);
+}
+
+/*! \return A target for ROCm */
+Target target_rocm() {
+    std::unordered_set<std::string> keys({ "rocm", "gpu" });
+    std::unordered_set<std::string> options;
+    return Target("rocm", kDLROCM, 256, 1, keys, options);
+}
+
+/*! \return A target for Metal */
+Target target_metal() {
+    std::unordered_set<std::string> keys({ "gpu" });
+    std::unordered_set<std::string> options;
+    return Target("metal", kDLMetal, 256, 1, keys, options);
+}
+
+/*! \return A target for rasp */
+Target target_rasp() {
+    std::unordered_set<std::string> keys({ "llvm", "cpu" });
+    std::unordered_set<std::string> options({
+        "-device=rasp",
+        "-mtriple=armv7l-none-linux-gnueabihf",
+        "-mcpu=cortex-a53",
+        "-mattr=+neon"
+    });
+    return Target("llvm", kDLCPU, 512, 1, keys, options);
+}
+
+/*! \return A target for stackvm */
+Target target_stackvm() {
+    std::unordered_set<std::string> keys({ "stackvm", "cpu" });
+    std::unordered_set<std::string> options;
+    return Target("stackvm", kDLCPU, 512, 1, keys, options);
+}
+
+
+bool LLVMEnabled() {
+    const runtime::PackedFunc* pf = runtime::Registry::Get("codegen.llvm_target_enabled");
```

Comment:
We have a function called target enabled, this function should not belong to the header. Move to cc